### PR TITLE
added public_name search function

### LINKED ISSUE #2082: Searching for the last name does not find an account
I just used the Liberapay search for "Owens". The profile I was looking for is this:

https://en.liberapay.com/doctormo/

and the name "Owens" is known to Liberapay as can be seen in the profile. Unfortunately, the Liberapay search only found unrelated accounts.

### PR DESCRIPTION
**Pull request by biskwikman**
closes #2082 

I basically mimicked the structure of the code for usernames, but pointed the new query to the `public_name` column of the `participants` table, it seems to work fine.

### CODE REVIEW

**Review by Changaco — CHANGES_REQUESTED** [review]
I think it's probably better to have a single query and unified results section for both usernames and public names.

Also, a new [trigram index](https://www.postgresql.org/docs/current/pgtrgm.html) is needed on the `public_name` column to speed up the search:

```sql
CREATE INDEX public_name_trgm_idx ON participants
    USING GIN (lower(public_name) gin_trgm_ops)
    WHERE status = 'active'
      AND public_name IS NOT null;
```

**Comment by biskwikman** [comment]
Should we include the username and public name under the avatar on all search results from now on? I think it would help with any confusion if an account has a similar username and public name.

**Comment by Changaco** [comment]
Yes, both names should be displayed for all results.

**Comment by biskwikman** [comment]
I've made it so the username and public name are returned in a single query, and both are displayed. I also added the index.

I'm not sure what word to use instead of "username" now. For example in the phrasing "Found a matching username". I was thinking of just using "user", but maybe "individual" would be more consistent.

File: www/search.spt [hunk 1]
```diff
@@ -15,12 +15,23 @@ if query:
...
+                 , id
               FROM participants
-             WHERE status = 'active'
-               AND NOT username like '~%%'
-               AND lower(username) %% lower(%(q)s)
+             WHERE lower(username) %% lower(%(q)s)
+                OR lower(public_name) %% lower(%(q)s)
+               AND status = 'active'
                AND hide_from_search = 0
+               AND NOT username like '~%%'
```
Comment (Changaco) [inline www/search.spt:34]: This is incorrect due to the order of precedence of the boolean operators. `… OR … AND …` is equivalent to `… OR ( … AND … )`, but here you wanted `( … OR … ) AND …`.

Also, you forgot to add `public_name IS NOT NULL`, so the query isn't using the new index.

```suggestion
             WHERE status = 'active'
               AND hide_from_search = 0
               AND ( NOT username like '~%%' AND lower(username) %% lower(%(q)s)
                     OR
                     public_name IS NOT NULL AND lower(public_name) %% lower(%(q)s)
                   )
```

File: www/search.spt [hunk 2]
```diff
@@ -15,12 +15,23 @@ if query:
...
+            SELECT username
+                 , public_name
+                 , avatar_url
+                 , CASE 
+                        WHEN public_name IS NULL 
+                             THEN similarity(username, %(q)s)
+                        WHEN similarity(username, %(q)s) >= similarity(public_name, %(q)s) 
+                             THEN similarity(username, %(q)s)
+                        ELSE similarity(public_name, %(q)s)
+                    END AS rank
```
Comment (Changaco) [inline www/search.spt:27]:
```suggestion
                 , greatest(
                       similarity(username, %(q)s),
                       similarity(public_name, %(q)s)
                   ) AS rank
```

**Review by Changaco — CHANGES_REQUESTED** [review]

**Comment by biskwikman** [comment]
thanks for the suggestions. I just wanted to check, the tests will have to be updated for this, right? It looks like two are failing now.

**Comment by Changaco** [comment]
The tests aren't failing with the latest commit. I think GitHub is still showing earlier test results because there's currently a merge conflict (because #2177 hasn't been deployed yet, because I've been waiting for new machine and human translations (https://github.com/liberapay/salon/issues/520)).

**Comment by Changaco** [comment]
Sorry I let this linger for two years. It's now merged (#2401) and will be deployed this week.

**Pull request closed without merging**